From 0b0bb8a3b93b7c4541e7bafec7fc3537020fb663 Mon Sep 17 00:00:00 2001
From: Yongjun Zhang <yzhang@cloudera.com>
Date: Tue, 3 Nov 2015 17:16:17 -0800
Subject: [PATCH 0969/2848] HDFS-9351. checkNNStartup() need to be called when
 fsck calls FSNamesystem.getSnapshottableDirs().
 (Xiao Chen via Yongjun Zhang)

(cherry picked from commit 194251c85250fcbe80a6ffee88b2cd4689334be3)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

Change-Id: I7302a6ee704af607032f802e50410b97d71f9cf1
---
 .../hadoop/hdfs/server/namenode/FSNamesystem.java  |   20 --------------------
 .../hadoop/hdfs/server/namenode/NamenodeFsck.java  |    9 ++++++++-
 2 files changed, 8 insertions(+), 21 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
index e24baea..4174dac 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java
@@ -8104,26 +8104,6 @@ public long getNumberOfSnapshottableDirs() {
     return list;
   }
 
-  /**
-   * Get the list of snapshottable directories.
-   * @return The list of all the current snapshottable directories
-   * @see #getSnapshottableDirListing()
-   * @throws IOException
-   */
-  List<String> getSnapshottableDirs() throws IOException {
-    List<String> snapshottableDirs = new ArrayList<String>();
-    final FSPermissionChecker pc = getPermissionChecker();
-    final String user = pc.isSuperUser() ? null : pc.getUser();
-    final SnapshottableDirectoryStatus[] snapDirs =
-        snapshotManager.getSnapshottableDirListing(user);
-    if (snapDirs != null) {
-      for (SnapshottableDirectoryStatus sds : snapDirs) {
-        snapshottableDirs.add(sds.getFullPath().toString());
-      }
-    }
-    return snapshottableDirs;
-  }
-
   @Override  //NameNodeMXBean
   public int getDistinctVersionCount() {
     return blockManager.getDatanodeManager().getDatanodesSoftwareVersions()
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java
index e578cdb..148593f 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java
@@ -57,6 +57,7 @@
 import org.apache.hadoop.hdfs.protocol.HdfsFileStatus;
 import org.apache.hadoop.hdfs.protocol.LocatedBlock;
 import org.apache.hadoop.hdfs.protocol.LocatedBlocks;
+import org.apache.hadoop.hdfs.protocol.SnapshottableDirectoryStatus;
 import org.apache.hadoop.hdfs.protocol.datatransfer.sasl.DataEncryptionKeyFactory;
 import org.apache.hadoop.hdfs.security.token.block.BlockTokenIdentifier;
 import org.apache.hadoop.hdfs.security.token.block.DataEncryptionKey;
@@ -322,7 +323,13 @@ public void fsck() {
       namenode.getNamesystem().logFsckEvent(path, remoteAddress);
 
       if (snapshottableDirs != null) {
-        snapshottableDirs = namenode.getNamesystem().getSnapshottableDirs();
+        SnapshottableDirectoryStatus[] snapshotDirs =
+            namenode.getRpcServer().getSnapshottableDirListing();
+        if (snapshotDirs != null) {
+          for (SnapshottableDirectoryStatus dir : snapshotDirs) {
+            snapshottableDirs.add(dir.getFullPath().toString());
+          }
+        }
       }
 
       final HdfsFileStatus file = namenode.getRpcServer().getFileInfo(path);
-- 
1.7.9.5

